### PR TITLE
`pyg-lib` integration in `HGTConv`

### PR DESCRIPTION
**Pull request by puririshi98**
ready🎉
```
data = FakeHeteroDataset(num_node_types=16, num_edge_types=128).data.to('cuda')
master branch iterative algorithm fwd pass time: 0.0014522361755371093
average fwd pass time(pyg-lib w/ torch<1.14): 0.0004720401763916016
average fwd pass time(pyg-lib w/ torch>=1.14): 0.00046291828155517576
```

3x speedup. As part of my blog I will do a more extensive benchmark to see how runtime for for-loop, segment_matmul, and group_matmul backends of HGT compare agaisnt eachother

smaller heterograph:
```
data = FakeHeteroDataset(num_node_types=8, num_edge_types=16).data.to('cuda')
average fwd pass time: 0.00020465850830078124
average fwd pass time(pyg-lib w/ torch<1.14): 0.00013236522674560546
average fwd pass time(pyg-lib w/ torch>=1.14): 0.00013113975524902343
```
1.6x speedup

scaling:
image

For interactive graph: https://github.com/puririshi98/rgcn_pyg_lib_forward_bench/blob/main/3d_plot.py

### CODE REVIEW

File: torch_geometric/nn/conv/hgt_conv.py [hunk 1]
```diff
@@ -142,29 +148,94 @@ def forward(
...
-            k = (k_dict[src_type].transpose(0, 1) @ a_rel).transpose(1, 0)
-
-            m_rel = self.m_rel[edge_type]
-            v = (v_dict[src_type].transpose(0, 1) @ m_rel).transpose(1, 0)
-
-            # propagate_type: (k: Tensor, q: Tensor, v: Tensor, rel: Tensor)
-            out = self.propagate(edge_index, k=k, q=q_dict[dst_type], v=v,
-                                 rel=self.p_rel[edge_type], size=None)
-            out_dict[dst_type].append(out)
+        if WITH_PYG_LIB:
```
Comment (rusty1s) [inline torch_geometric/nn/conv/hgt_conv.py:151]: needs to be `torch_geometric.typing.WITH_PYG_LIB` to enable TorchScript support (don't ask me why :D)

File: torch_geometric/nn/conv/hgt_conv.py [hunk 2]
```diff
@@ -142,29 +148,94 @@ def forward(
...
+            # parralelize over node-types
+            node_types, xs = x_dict.keys(), x_dict.values()
+
+            # compute K
+            k_wts = [self.k_lin[node_type].weight for node_type in node_types]
+            k_biases = [self.k_lin[node_type].bias for node_type in node_types]
+            k_list = pyg_lib.ops.grouped_matmul(inputs=xs, others=k_wts,
```
Comment (rusty1s) [inline torch_geometric/nn/conv/hgt_conv.py:158]: I would prefer to use `segment_matmul` here and first concatenate `x_dict` together. Otherwise, we are restricting ourselves to PyTorch 1.14/2.0 here which is not even released.

File: torch_geometric/nn/conv/hgt_conv.py (hunk 1 above)
Comment (rusty1s) [inline torch_geometric/nn/conv/hgt_conv.py:151]: I am personally fine with making `PYG_LIB` mandatory here - I guess this avoids a lot of diverging code paths.

File: torch_geometric/nn/conv/hgt_conv.py [hunk 3]
```diff
@@ -142,29 +148,94 @@ def forward(
...
+            v_outs = [
+                v_o_i.transpose(1, 0)
+                for v_o_i in pyg_lib.ops.grouped_matmul(v_ins, m_rels)
+            ]
+            del v_ins
+            for i, dst_type in enumerate(dst_types):
+                out = self.propagate(e_idxs[i], k=k_outs[i],
+                                     q=q_dict[dst_type], v=v_outs[i],
+                                     rel=self.p_rel[edge_types[i]], size=None)
+                out_dict[dst_type].append(out)
```
Comment (rusty1s) [inline torch_geometric/nn/conv/hgt_conv.py:214]: I think we can only propagate once here if we stack `edge_index_dict` into a single `edge_index` matrix.

File: torch_geometric/nn/conv/hgt_conv.py (hunk 1 above)
Comment (puririshi98) [reply torch_geometric/nn/conv/hgt_conv.py:151]: 😝

**Comment by puririshi98** [comment]
still need to test and benchmark it but done w/ an initial draft

**Comment by puririshi98** [comment]
@rusty1s [this PR ](https://github.com/pyg-team/pyg-lib/pull/183)to pyg-lib is needed for torch >= 1.14
I checked locally that once I build pyg-lib w/ that PR included, the grouped matmul and segment matmul pathways both work except i get the wierd numerical mismatch between the results of using coo edge indices vs csr indices:


```
        for node_type in out_dict1.keys():
>           assert torch.allclose(out_dict1[node_type], out_dict2[node_type],
                                  atol=1e-6)
E           assert False
E            +  where False = <built-in method allclose of type object at 0x7f5817c2abc0>(tensor([[-0.0835, -0.6736,  0.5235,  0.2286, -0.0260,  0.3432, -0.3394, -0.0610,\n          0.6782,  0.3317,  0.3618,  0.3018, -0.1857,  0.1624, -0.4920,  0.0622],\n        [-0.0600,  0.0160,  0.2255,  0.0513,  0.0783, -0.0502,  0.2167,  0.1157,\n          0.0927,  0.0175, -0.0928,  0.0178, -0.0471,  0.2207, -0.1846, -0.3963],\n        [-0.1174,  0.0345,  0.3031,  0.0869,  0.3046,  0.0708, -0.4384,  0.0139,\n          0.0960,  0.5766,  0.0997, -0.3248, -0.3266,  0.0738, -0.2258,  0.3176],\n        [ 0.2621,  0.1540,  0.2881,  0.1190, -0.1480, -0.0665,  0.3536, -0.0015,\n         -0.1051, -0.0060, -0.2733,  0.3685, -0.3494,  0.3862, -0.1167,  0.1122]],\n       grad_fn=<AddBackward0>), tensor([[-0.1288, -0.4715,  0.6094,  0.0883,  0.0351,  0.2979, -0.4755,  0.0921,\n          0.7698,  0.1106,  0.4882,  0.2792, -0.3037,  0.3606, -0.6694,  0.0487],\n        [-0.0075,  0.2320,  0.3542, -0.1559,  0.1909, -0.0801,  0.0948,  0.3497,\n          0.1253, -0.2684,  0.0116,  0.2078, -0.1117,  0.3709, -0.3056, -0.4902],\n        [-0.0718,  0.1115,  0.4493,  0.0243,  0.3455,  0.1374, -0.3731,  0.0014,\n          0.1058,  0.4818,  0.1863, -0.1104, -0.2884,  0.1684, -0.2276,  0.3486],\n        [ 0.2621,  0.1540,  0.2881,  0.1190, -0.1480, -0.0665,  0.3536, -0.0015,\n         -0.1051, -0.0060, -0.2733,  0.3685, -0.3494,  0.3862, -0.1167,  0.1122]],\n       grad_fn=<AddBackward0>), atol=1e-06)
E            +    where <built-in method allclose of type object at 0x7f5817c2abc0> = torch.allclose
```

**Comment by puririshi98** [comment]
when i try to run the following benchmark:
```
from torch_geometric.nn.conv import HGTConv
from torch_geometric.datasets import FakeHeteroDataset
import time
data = FakeHeteroDataset(num_node_types=16, num_edge_types=128).data.to('cuda')
net = HGTConv(-1, 32, data.metadata()).to('cuda')
x_dict = data.collect('x')
edge_index_dict = data.collect('edge_index')
for i in range(60):
  if i > 9:
    since = time.time()
  net(x_dict, edge_index_dict)
print("average fwd pass time:", (time.time()-since)/50.0)
```

i get the following error: 
```
Traceback (most recent call last):
  File "x.py", line 11, in <module>
    net(x_dict, edge_index_dict)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1480, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/torch_geometric-2.3.0-py3.8.egg/torch_geometric/nn/conv/hgt_conv.py", line 395, in forward
    out = self.propagate(e_idx, k=k_out, q=q, v=v_out, rel=p, size=None)
  File "/usr/local/lib/python3.8/dist-packages/torch_geometric-2.3.0-py3.8.egg/torch_geometric/nn/conv/message_passing.py", line 432, in propagate
    coll_dict = self.__collect__(self.__user_args__, edge_index,
  File "/usr/local/lib/python3.8/dist-packages/torch_geometric-2.3.0-py3.8.egg/torch_geometric/nn/conv/message_passing.py", line 304, in __collect__
    data = self.__lift__(data, edge_index, dim)
  File "/usr/local/lib/python3.8/dist-packages/torch_geometric-2.3.0-py3.8.egg/torch_geometric/nn/conv/message_passing.py", line 242, in __lift__
    raise IndexError(
IndexError: Encountered an index error. Please ensure that all indices in 'edge_index' point to valid indices in the interval [0, 130169] (got interval [169034, 259441])
```

to me this implies theres a problem w/ me going from edge_index_dict to the single edge index. will investigate further. lmk if anything is obvious to you

**Comment by puririshi98** [comment]
copied your code for making edgeindexdict into and edge index tensor and i still get numerical mismatches btwn adj_t and edge_index use

**Comment by puririshi98** [comment]
debugging prints:
```
        # propagate
        print("\n"*2)
        print("k[0,0] =", k_out[0,0])
        print("q[0,0] =", q[0,0])
        print("v[0,0] =", v_out[0,0])
        print("p[0,0] =", p)
        out = self.propagate(e_idx, k=k_out, q=q, v=v_out, rel=p, size=None)
        print("propogate_out[0,0] =", out[0,0])
```
output:
```
coo format:
k[0,0] = tensor([-0.4085, -0.4570, -0.2870,  0.5745,  0.3404,  0.4127,  0.6706,  0.2871],
       grad_fn=<SelectBackward0>)
q[0,0] = tensor([ 0.0614, -0.6190, -1.3580, -0.6144,  0.4729,  0.7143,  0.2142,  0.6712],
       grad_fn=<SelectBackward0>)
v[0,0] = tensor([ 0.1358,  0.2227, -0.3013, -0.0545, -1.0439,  0.2239,  1.1845, -0.0182],
       grad_fn=<SelectBackward0>)
p[0,0] = tensor([2., 2.], grad_fn=<ViewBackward0>)
propogate_out[0,0] = tensor(0.3956, grad_fn=<SelectBackward0>)


csr format:
k[0,0] = tensor([-0.4085, -0.4570, -0.2870,  0.5745,  0.3404,  0.4127,  0.6706,  0.2871],
       grad_fn=<SelectBackward0>)
q[0,0] = tensor([ 0.0614, -0.6190, -1.3580, -0.6144,  0.4729,  0.7143,  0.2142,  0.6712],
       grad_fn=<SelectBackward0>)
v[0,0] = tensor([ 0.1358,  0.2227, -0.3013, -0.0545, -1.0439,  0.2239,  1.1845, -0.0182],
       grad_fn=<SelectBackward0>)
p[0,0] = tensor([2., 2.], grad_fn=<ViewBackward0>)
propogate_out[0,0] = tensor(-0.2069, grad_fn=<SelectBackward0>)
```

notice that the inputs to propagate are identical but the outputs are not. Somehow [converting the edge_indices from coo to csr ](https://github.com/pyg-team/pytorch_geometric/blob/hgtconv_pyglib/torch_geometric/nn/conv/hgt_conv.py#L367-L385)changes the results of propogate for identical inputs.

**Comment by puririshi98** [comment]
I have isolated the reason for the test failures where HGT outputs don't match between coo and csr edge_index representation. Notice in the debug output below that k, q, v, and p are identical between both runs, but the edge indices don't match which explains why the output of propagate does not match. There seems to be an issue in converting csr to coo and then back to csr. @rusty1s could you please take a look and see if I am doing anything noticeably wrong?


[debugging prints](https://github.com/pyg-team/pytorch_geometric/blob/hgtconv_pyglib/torch_geometric/nn/conv/hgt_conv.py#L362-L403):

output:
```
coo format:
e_idx[('author', 'writes', 'paper')] = tensor([[1, 2, 0, 1, 1, 1, 0, 3, 1, 3, 2, 1, 0, 3, 1, 2, 1, 3, 0, 1],
        [1, 1, 3, 4, 2, 5, 3, 5, 2, 2, 0, 1, 1, 0, 5, 1, 4, 2, 2, 4]])
e_idx[('paper', 'written_by', 'author')] = tensor([[1, 1, 3, 4, 2, 5, 3, 5, 2, 2, 0, 1, 1, 0, 5, 1, 4, 2, 2, 4],
        [1, 2, 0, 1, 1, 1, 0, 3, 1, 3, 2, 1, 0, 3, 1, 2, 1, 3, 0, 1]])
k[0,0] = tensor([0.6495, 0.3871, 0.5111, 0.0973, 0.1445, 0.5161, 0.4522, 0.1621],
       grad_fn=<SelectBackward0>)
k.shape = torch.Size([10, 2, 8])
q[0,0] = tensor([ 0.1918, -0.7619,  0.7618,  0.1111,  0.4187,  1.5218,  0.3339, -0.9197],
       grad_fn=<SelectBackward0>)
q.shape = torch.Size([10, 2, 8])
v[0,0] = tensor([-0.1232, -0.0434, -0.1867, -0.7753,  0.3347, -0.1346,  0.1867, -0.2690],
       grad_fn=<SelectBackward0>)
v.shape = torch.Size([10, 2, 8])
p[0,0] = tensor([2., 2.], grad_fn=<ViewBackward0>)
e_idx = tensor([[1, 2, 0, 1, 1, 1, 0, 3, 1, 3, 2, 1, 0, 3, 1, 2, 1, 3, 0, 1, 5, 5, 7, 8,
         6, 9, 7, 9, 6, 6, 4, 5, 5, 4, 9, 5, 8, 6, 6, 8],
        [5, 5, 7, 8, 6, 9, 7, 9, 6, 6, 4, 5, 5, 4, 9, 5, 8, 6, 6, 8, 1, 2, 0, 1,
         1, 1, 0, 3, 1, 3, 2, 1, 0, 3, 1, 2, 1, 3, 0, 1]])
propagate_out[0,0] = tensor(-0.0425, grad_fn=<SelectBackward0>)


csr format (converting to coo and then back to csr under the hood to combine adj_t_dict into adj_t):
e_idx[('author', 'writes', 'paper')] = tensor([[0, 0, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 3, 3, 4, 4, 4, 5, 5, 5],
        [2, 3, 0, 1, 1, 2, 2, 0, 1, 1, 3, 3, 0, 0, 1, 1, 1, 1, 1, 3]])
e_idx[('paper', 'written_by', 'author')] = tensor([[0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 3, 3, 3, 3],
        [1, 2, 3, 3, 1, 1, 2, 2, 4, 4, 4, 5, 5, 0, 1, 1, 0, 2, 2, 5]])
k[0,0] = tensor([0.6495, 0.3871, 0.5111, 0.0973, 0.1445, 0.5161, 0.4522, 0.1621],
       grad_fn=<SelectBackward0>)
k.shape = torch.Size([10, 2, 8])
q[0,0] = tensor([ 0.1918, -0.7619,  0.7618,  0.1111,  0.4187,  1.5218,  0.3339, -0.9197],
       grad_fn=<SelectBackward0>)
q.shape = torch.Size([10, 2, 8])
v[0,0] = tensor([-0.1232, -0.0434, -0.1867, -0.7753,  0.3347, -0.1346,  0.1867, -0.2690],
       grad_fn=<SelectBackward0>)
v.shape = torch.Size([10, 2, 8])
p[0,0] = tensor([2., 2.], grad_fn=<ViewBackward0>)
e_idx = tensor([[0, 0, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 3, 3, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5,
         5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 7, 7, 7, 7],
        [6, 7, 4, 5, 5, 6, 6, 4, 5, 5, 7, 7, 4, 4, 1, 2, 3, 3, 5, 5, 5, 1, 1, 2,
         2, 4, 4, 4, 5, 5, 5, 5, 7, 0, 1, 1, 0, 2, 2, 5]])
propagate_out[0,0] = tensor(-0.3010, grad_fn=<SelectBackward0>)
```

**Comment by puririshi98** [comment]
figured it out lolz, the src n dst were backwards but since the conversion sorts it was hard to realize that at first glance of the edge indices

File: torch_geometric/nn/conv/hgt_conv.py [hunk 4]
```diff
@@ -139,38 +174,211 @@ def forward(
...
+        q_biases = [
+            self.q_lin[node_type].bias for node_type in self.node_types
+        ]
+        v_wts = [
+            self.v_lin[node_type].weight.T for node_type in self.node_types
+        ]
+        v_biases = [
+            self.v_lin[node_type].bias for node_type in self.node_types
+        ]
+        out_dict = {node_type: [] for node_type in self.node_types}
```
Comment (rusty1s) [inline torch_geometric/nn/conv/hgt_conv.py:227]: `defaultdict(list)`

File: torch_geometric/nn/conv/hgt_conv.py [hunk 5]
```diff
@@ -139,38 +174,211 @@ def forward(
...
+            ptr = torch.tensor(ptr).to(x.device)
+            k_wt = torch.stack(k_wts)
+            k_bias = torch.stack(k_biases)
+            q_wt = torch.stack(q_wts)
+            q_bias = torch.stack(q_biases)
+            v_wt = torch.stack(v_wts)
+            v_bias = torch.stack(v_biases)
+
+        # compute K, Q, V over node-types
+        if self.use_gmm:
```
Comment (rusty1s) [inline torch_geometric/nn/conv/hgt_conv.py:244]: Can we add a `pragma: no cover` comment to all `use_gmm` code paths for now?

File: torch_geometric/nn/conv/hgt_conv.py [hunk 6]
```diff
@@ -139,38 +174,211 @@ def forward(
...
+            # initialize lazy params
+            for node_type, x_n in x_dict.items():
+                self.k_lin[node_type].initialize_parameters(None, x_n)
+                self.q_lin[node_type].initialize_parameters(None, x_n)
+                self.v_lin[node_type].initialize_parameters(None, x_n)
+            self.infer_shapes = False
         H, D = self.heads, self.out_channels // self.heads
 
         k_dict, q_dict, v_dict, out_dict = {}, {}, {}, {}
+        # parralelize over node-types
```
Comment (rusty1s) [inline torch_geometric/nn/conv/hgt_conv.py:208]: typo.

Comment (rusty1s) [inline torch_geometric/nn/conv/hgt_conv.py:208]: I feel there is a lot of repetitive logic here. Isn't that something that `HeteroLinear` should be handling (or `HeteroDictLinear`?

File: torch_geometric/nn/conv/hgt_conv.py [hunk 7]
```diff
@@ -139,38 +174,211 @@ def forward(
...
+            }
+            v = pyg_lib.ops.segment_matmul(inputs=x, ptr=ptr, other=v_wt,
+                                           bias=v_bias)
+            v_dict = {
+                node_type: v[ptr[i]:ptr[i + 1]].view(-1, H, D)
+                for i, node_type in enumerate(self.node_types)
+            }
+
+        # parallelize over edge-types
+        src_types = [edge_type[0] for edge_type in self.edge_types]
```
Comment (rusty1s) [inline torch_geometric/nn/conv/hgt_conv.py:287]: Pre-compute in `__init__`?

File: torch_geometric/nn/conv/hgt_conv.py [hunk 8]
```diff
@@ -139,38 +174,211 @@ def forward(
...
+            trans_ptr = [count]
+            for src_type in src_types:
+                k_src = k_dict[src_type]
+                v_src = v_dict[src_type]
+                k_ins.append(k_src.view(-1, D))
+                v_ins.append(v_src.view(-1, D))
+                for i in range(H):
+                    count += k_src.size(0)
+                    trans_ptr.append(count)
+            trans_ptr = torch.tensor(trans_ptr)
```
Comment (rusty1s) [inline torch_geometric/nn/conv/hgt_conv.py:330]: device?

**Review by rusty1s — COMMENTED** [review]
Thanks for the PR, sorry that it took me so long to take a look.

I think overall the logic makes sense to me, but it is a bit hard to parse, mostly due to all the different code paths. In particular, I am not entirely sure why we need to merge this all into `HGTConv`. Can't we out-source some utilities and `grouped_matmul` behavior into separate functions/modules?

File: torch_geometric/nn/conv/hgt_conv.py (hunk 8 above)
Comment (puririshi98) [reply torch_geometric/nn/conv/hgt_conv.py:330]: > For best performance, given as a CPU tensor.
don't we recommend cpu tensors for segment matmul?

File: torch_geometric/nn/conv/hgt_conv.py [hunk 9]
```diff
@@ -139,38 +174,211 @@ def forward(
...
+                    self.max_channels = self.dims.max()
+                    input_to_init_w = xs[self.dims.argmax()]
+                    for node_type in x_dict.keys():
+                        self.k_lin[node_type].initialize_parameters(
+                            None, input_to_init_w)
+                        self.q_lin[node_type].initialize_parameters(
+                            None, input_to_init_w)
+                        self.v_lin[node_type].initialize_parameters(
+                            None, input_to_init_w)
+                    self.infer_shapes = False
```
[the remark this answers is not in the document]
Comment (puririshi98) [reply torch_geometric/nn/conv/hgt_conv.py]: done

File: torch_geometric/nn/conv/hgt_conv.py [hunk 10]
```diff
@@ -139,38 +174,211 @@ def forward(
...
+            q_list.append(q_dict[dst_type])
+            p_rels.append(self.p_rel['__'.join(e_type)].view(-1, 1))
+
+        q = torch.cat(q_list)
+        p = group(p_rels, self.group).view(-1)
+        e_idx = torch.cat(edge_index_list, dim=1)
+        if convert:
+            # convert back to CSR
+            e_idx = SparseTensor(row=e_idx[0], col=e_idx[1],
+                                 sparse_sizes=(k_out.size(0), k_out.size(0)))
```
[the remark this answers is not in the document]
Comment (puririshi98) [reply torch_geometric/nn/conv/hgt_conv.py]: yes will do

File: torch_geometric/nn/conv/hgt_conv.py (hunk 6 above)
Comment (puririshi98) [reply torch_geometric/nn/conv/hgt_conv.py]: yes will do

File: torch_geometric/nn/conv/hgt_conv.py (hunk 8 above)
Comment (rusty1s) [reply torch_geometric/nn/conv/hgt_conv.py:316]: Oh, you are right. Thanks for clarifying.

File: torch_geometric/nn/conv/hgt_conv.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (puririshi98) [reply torch_geometric/nn/conv/hgt_conv.py]: cleaned up now

File: torch_geometric/nn/conv/hgt_conv.py (hunk 6 above)
Comment (puririshi98) [reply torch_geometric/nn/conv/hgt_conv.py]: done

Comment (puririshi98) [reply torch_geometric/nn/conv/hgt_conv.py]: done

**Comment by puririshi98** [comment]
finished a major cleanup thanks to the review @rusty1s. In doing so I realized I should be parralelizing a_lin as well which further improved the biggest speedboost to 3.6x, see [blog](https://docs.google.com/document/d/1uRB4IpjL5DlS3YG19eNnX9zSvGzgK7JdNX10aXjdDy8/edit?usp=sharing) for updated benchmark plot

**Review by wsad1 — COMMENTED** [review]
Left some initial comments. Will take a look at it again.

File: torch_geometric/nn/conv/hgt_conv.py [hunk 11]
```diff
@@ -143,40 +178,123 @@ def forward(
...
+        }
+
+        # parallelize over edge-types
+        a_rels = [
+            self.a_rel['__'.join(edge_type)] for edge_type in self.edge_types
+        ]
+        m_rels = [
+            self.m_rel['__'.join(edge_type)] for edge_type in self.edge_types
+        ]
+        if self.use_gmm:  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (puririshi98) [reply torch_geometric/nn/conv/hgt_conv.py:206]: valid point, will update

File: torch_geometric/nn/conv/hgt_conv.py [hunk 12]
```diff
@@ -187,7 +305,6 @@ def forward(
     def message(self, k_j: Tensor, q_i: Tensor, v_j: Tensor, rel: Tensor,
                 index: Tensor, ptr: Optional[Tensor],
                 size_i: Optional[int]) -> Tensor:
-
```
[the remark this answers is not in the document]
Comment (puririshi98) [reply torch_geometric/nn/conv/hgt_conv.py:190]: hmmm im not 100% sure, @rusty1s can u advise on this? i did my best to match up the shapes/dimensions and have the hgt CI tests passes.

[the remark this answers is not in the document]
Comment (rusty1s) [reply torch_geometric/nn/conv/hgt_conv.py:190]: I think the two implementations are definitely different, but I believe we are fixing the original implementation here (since `HGTConv` is indeed designed to share message passing over all edge types AFAIK).

File: torch_geometric/nn/conv/hgt_conv.py [hunk 13]
```diff
@@ -143,40 +178,105 @@ def forward(
...
+            # (TODO) Add support for SparseTensor w/o converting?
+            convert = isinstance(indices, SparseTensor)
+            if convert:
+                # convert to COO
+                dst, src, _ = indices.coo()
+                indices = torch.cat((src.view(1, -1), dst.view(1, -1)))
+            edge_index_list.append(
+                offset_edge_idx(node_slices, e_type, indices))
+            q_list.append(q_dict[e_type[-1]])
+            p_rels.append(self.p_rel['__'.join(e_type)].view(-1, 1))
```
Comment (wsad1) [inline torch_geometric/nn/conv/hgt_conv.py:248]: I have a concern with `node_slices` and how the `offset_edge_idx` is being applied. I thought it would be best to share an example to illustrate the issue I see: 

I added a few print stmts to the code (we'll remove them before merging) and ran the below example:
```
import torch
from torch_geometric.nn import HGTConv

x_dict = {
    'author': torch.randn(4, 16),
    'paper': torch.randn(6, 16),
}

index1 = torch.randint(0, 4, (20, ), dtype=torch.long)
index2 = torch.randint(0, 6, (20, ), dtype=torch.long)

edge_index_dict = {
    ('author', 'writes', 'paper'): torch.stack([index1, index2]),
    ('paper', 'written_by', 'author'): torch.stack([index2, index1]),
    ('paper', 'liked_by', 'author'): torch.stack([index2, index1]),
}
metadata = (list(x_dict.keys()), list(edge_index_dict.keys()))
conv = HGTConv(16, 16, metadata, heads=2)
out_dict1 = conv(x_dict, edge_index_dict)
```
The output is as follows:
```
Trans torch.Size([7]) tensor([ 0,  4,  8, 14, 20, 26, 32])
K_out torch.Size([16, 2, 8])
node_slices {'author': (0, 4), 'paper': (4, 8)}
Q torch.Size([14, 2, 8])
```
In the above case the `node_slices` I believe are for `k_out` but since `k_out` has node types repeated, since `src_types` has repeated node types, I don't see how these `node_slices` can be used to offset the `edge_index` appropriately. 

Again I might be missing something, since the code is kinda complex.

**Review by wsad1 — COMMENTED** [review]
Thanks @puririshi98 for the updates. 
Left another set of reviews. Of which only the last and longer one is a concern for me.

File: torch_geometric/nn/conv/hgt_conv.py (hunk 13 above)
Comment (puririshi98) [reply torch_geometric/nn/conv/hgt_conv.py:247]: this code for the node_slices is exactly copied from `to_homogeneous_edge_index` which was written by @rusty1s 

from my understanding this code has been working for cases even when there are multiple of the same source types. the node slices are applied as offsets to the edge indices w/:
```
def offset_edge_idx(
    node_slices: Dict[NodeType, Tuple[int, int]],
    edge_type: EdgeType,
    edge_index: Tensor,
) -> Tuple[int, int]:
    src, _, dst = edge_type
    offset = [[node_slices[src][0]], [node_slices[dst][0]]]
    offset = torch.tensor(offset, device=edge_index.device)
    return edge_index + offset
```

im not sure if i see any issue. @rusty1s plz advise

Comment (rusty1s) [reply torch_geometric/nn/conv/hgt_conv.py:247]: Also discussed with @wsad1. I haven't looked to closely yet, but I think he is right. `K_out` seems to be of shape `[16, 2, 8]`, where `16 == 4(author)+6(paper)+6(paper)`, so we cannot use `node_slices` here. We would need an edge-type dependent `node_slices` vector instead.

Comment (wsad1) [reply torch_geometric/nn/conv/hgt_conv.py:247]: Thanks for explaining that. 
Yeah we need `node_slices` per edge type, further the offset for edge_index[0] should depend on slices in `k_out` and for edge_index[1] it should depend on slices in `q`. Happy to update this if that helps.

Comment (puririshi98) [reply torch_geometric/nn/conv/hgt_conv.py:247]: > Happy to update the PR, if that helps.

@wsad1  that would be appreciated if possible. thanks!

File: torch_geometric/nn/conv/hgt_conv.py [hunk 14]
```diff
@@ -143,40 +177,105 @@ def forward(
...
+
+        q = torch.cat(q_list)
+        print('Q', q.shape)
+        p = group(p_rels, self.group).view(-1)
```
Comment (wsad1) [inline torch_geometric/nn/conv/hgt_conv.py:251]:
```suggestion
        # Concatenate all dst node tensors.
        q_list = []
        dst_offset = {}
        offset = 0
        for node_type in self.node_types:
            q_list.append(q_dict[node_type])
            dst_offset[node_type] = offset
            offset += q_dict[node_type].size(0)
        q = torch.cat(q_list)

        # list of edges and edge tensors.
        edge_index_list = []
        edge_src_offset = 0
        p_rels = []


        # list of src node tensors for each edge
        k_ins = []
        v_ins = []
        count = 0
        trans_ptr = [count]


        for edge_type in self.edge_types:
            src_type, _, dst_type = edge_type

            # Collect src node tesnors.
            k_src = k_dict[src_type]
            v_src = v_dict[src_type]
            k_ins.append(k_src.view(-1, D))
            v_ins.append(v_src.view(-1, D))
            for _ in range(H):
                count += k_src.size(0)
                trans_ptr.append(count)

            # Collect edges and edge tensors.
            p_rels.append(self.p_rel['__'.join(edge_type)].view(-1, 1))
            indices = edge_index_dict[edge_type]
            # (TODO) Add support for SparseTensor w/o converting?
            convert = isinstance(indices, SparseTensor)
            if convert:
                # convert to COO
                dst, src, _ = indices.coo()
                indices = torch.cat((src.view(1, -1), dst.view(1, -1)))
            # Add offset to edge indices. Offset needs to be added as
            # the src and dst node tensors are concatenated.
            edge_index_list.append(offset_edge(indices, (edge_src_offset, dst_offset[dst_type])))

            # Update offsets.
            edge_src_offset += k_src.size(0)

        # Concatenate all src node tensors.
        trans_ptr = torch.tensor(trans_ptr)
        a_rel, m_rel = torch.cat(a_rels), torch.cat(m_rels)
        k_out = pyg_lib.ops.segment_matmul(torch.cat(k_ins), trans_ptr,
                                           a_rel).view(-1, H, D)
        v_out = pyg_lib.ops.segment_matmul(torch.cat(v_ins), trans_ptr,
                                           m_rel).view(-1, H, D)

        p = group(p_rels, self.group).view(-1)
```
Here's the suggested changes as we discussed. 
I am still not sure about how we compute `p`. But i'll leave that for later.

Comment (puririshi98) [reply torch_geometric/nn/conv/hgt_conv.py:251]: thanks! it makes alot more sense to me now seeing the code :)

**Comment by puririshi98** [comment]
thanks for the reviews, lmk if anything else needed to merge @wsad1 @rusty1s

File: torch_geometric/data/hetero_data.py [hunk 15]
```diff
@@ -976,35 +976,55 @@ def get_all_edge_attrs(self) -> List[EdgeAttr]:
...
+        node_slices[node_type] = (cumsum, cumsum + num_nodes)
+        cumsum += num_nodes
+    return node_slices
+
+
+def offset_edge_idx(
+    node_slices: Dict[NodeType, Tuple[int, int]],
+    edge_type: EdgeType,
+    edge_index: Tensor,
+) -> Tuple[int, int]:
```
Comment (wsad1) [inline torch_geometric/data/hetero_data.py:995]:
```suggestion
) -> Tensor:
```

File: torch_geometric/nn/conv/hgt_conv.py [hunk 16]
```diff
@@ -78,23 +86,45 @@ def __init__(
 
         if not isinstance(in_channels, dict):
             in_channels = {node_type: in_channels for node_type in metadata[0]}
-
+        self.use_gmm = grouped_matmul_avail()
         self.in_channels = in_channels
         self.out_channels = out_channels
         self.heads = heads
         self.group = group
```
Comment (wsad1) [inline torch_geometric/nn/conv/hgt_conv.py:93]: Lets drop the group argument. Its not really used to group node tensors any more.

File: torch_geometric/nn/conv/hgt_conv.py [hunk 17]
```diff
@@ -143,40 +173,118 @@ def forward(
...
+
+        # Concatenate all src node tensors.
+        trans_ptr = torch.tensor(trans_ptr)
+        a_rel, m_rel = torch.cat(a_rels), torch.cat(m_rels)
+        k_out = pyg_lib.ops.segment_matmul(torch.cat(k_ins), trans_ptr,
+                                           a_rel).view(-1, H, D)
+        v_out = pyg_lib.ops.segment_matmul(torch.cat(v_ins), trans_ptr,
+                                           m_rel).view(-1, H, D)
+
+        p = group(p_rels, self.group).view(-1)
```
Comment (wsad1) [inline torch_geometric/nn/conv/hgt_conv.py:261]:
```suggestion
            indices = edge_index_dict[edge_type]
            # (TODO) Add support for SparseTensor w/o converting?
            convert = isinstance(indices, SparseTensor)
            if convert:
                # Convert to COO
                dst, src, _ = indices.coo()
                indices = torch.cat((src.view(1, -1), dst.view(1, -1)))
            num_edges = indices.size(1)
            p_rels.append(self.p_rel['__'.join(edge_type)].view(1, H).expand(num_edges, -1))
            # Add offset to edge indices. Offset needs to be applied as
            # the src and dst node tensors are concatenated.
            indices[0] += src_offset
            indices[1] += dst_offset[dst_type]
            edge_index_list.append(indices)

            # Update offset.
            src_offset += k_src.size(0)

        # Concatenate all src node tensors.
        trans_ptr = torch.tensor(trans_ptr)
        a_rel, m_rel = torch.cat(a_rels), torch.cat(m_rels)
        k_out = pyg_lib.ops.segment_matmul(torch.cat(k_ins), trans_ptr,
                                           a_rel).view(-1, H, D)
        v_out = pyg_lib.ops.segment_matmul(torch.cat(v_ins), trans_ptr,
                                           m_rel).view(-1, H, D)

        # Concatenate all edges and edge tensors.
        p = torch.cat(p_rels, dim = 0)
```
Each edge type should use a different `p`/`rel`. Updating how p is created to reflect that.

**Review by wsad1 — COMMENTED** [review]
This is looking ready to merge now mostly.
Left a last comment on how `p` should be computed, other than that the code looks correct. We can refactor later to enhance readability.

**Review by wsad1 — APPROVED** [review]
Thanks @puririshi98 for all the work. 
@rusty1s appreciate a quick look.

**Comment by puririshi98** [comment]
unfortunately, at the moment, even if i set all params to zero for the new hgt vs old hgt, the outputs dont match
test: https://github.com/puririshi98/rgcn_pyg_lib_forward_bench/blob/main/hgt_correctness.py

**Comment by puririshi98** [comment]
basic correctness passes: https://github.com/puririshi98/rgcn_pyg_lib_forward_bench/blob/main/hgt_correctness.py
@rusty1s merge when u can :)
